relay-replays/src/recording.rs

```diff
@@ -68,31 +91,58 @@ impl From<serde_json::Error> for ParseRecordingError {
     }
 }
 
-static STRING_STATE: Lazy<ProcessingState> = Lazy::new(|| {
-    ProcessingState::root().enter_static(
-        "",
-        Some(Cow::Owned(FieldAttrs::new().pii(Pii::True))),
-        Some(ValueType::String),
-    )
-});
-
 /// The [`Transform`] implementation for data scrubbing.
 ///
 /// This is used by [`EventStreamVisitor`] and [`ScrubbedValue`] to scrub recording events.
 struct ScrubberTransform<'a> {
+    /// PII processors that are applied one by one on each value.
     processor1: Option<PiiProcessor<'a>>,
     processor2: Option<PiiProcessor<'a>>,
+    /// The state encoding the current path, which is fed by `push_path` and `pop_path`.
+    state: ProcessingState<'a>,
+    /// The current path. This is redundant with `state`, which also contains the full path,
+    /// but easier to match on.
+    path: Vec<String>,
+}
+
+impl ScrubberTransform<'_> {
+    fn reset(&mut self) {
+        self.state = ProcessingState::new_root(None, None);
+        self.path.clear();
+    }
 }
 
-impl Transform for &'_ mut ScrubberTransform<'_> {
+impl<'de> Transform<'de> for &'_ mut ScrubberTransform<'_> {
+    fn push_path(&mut self, key: &'de str) {
+        self.path.push(key.to_owned());
+        let field_attrs = if scrub_at_path(&self.path) {
+            &FIELD_ATTRS_PII_TRUE
+        } else {
+            &FIELD_ATTRS_PII_FALSE
+        };
+
+        self.state = std::mem::take(&mut self.state).enter_owned(
+            key.to_owned(),
+            Some(Cow::Borrowed(field_attrs)),
+            Some(ValueType::String), // Pretend everything is a string.
+        )
+    }
+
+    fn pop_path(&mut self) {
+        if let Ok(Some(parent)) = std::mem::take(&mut self.state).try_into_parent() {
+            self.state = parent;
+        }
+        self.path.pop();
```

Comment:
`pop` returns `Option<T>`. Should we log an error if `pop` returns `None`? It means we tried to walk up to a parent while being in the root, which should never happen.